handlers/pod.py

```diff
@@ -7,19 +7,36 @@
 client = DatastoreClient()
 
 
-def add_update_pod(key, pod):
+def add_update_pod(key, pod, cache):
     """
     Takes a new or updated pod from the Kubernetes API and
     creates the corresponding Calico configuration.
     """
     _log.debug("Adding new pod: %s", key)
 
-    # Get the Calico endpoint.  This may or may not have already been
-    # created by the CNI plugin.  If it hasn't been created, we need to
-    # wait until is has before we can do any meaningful work.
+    # Get Kubernetes labels.
+    labels = pod["metadata"].get("labels", {})
+    _log.debug("Pod '%s' has labels: %s", key, labels)
+
+    # Extract information.
     namespace = pod["metadata"]["namespace"]
     name = pod["metadata"]["name"]
     workload_id = "%s.%s" % (namespace, name)
+
+    # Add a special label for the Kubernetes namespace.  This is used
+    # by selector-based policies to select all pods in a given namespace.
+    labels[K8S_NAMESPACE_LABEL] = namespace
+
+    # Check if the labels have changed - if they haven't, just return.
+    old_labels = cache.get(workload_id, {})
```

Comment:
Hmm, this is wrong.  I shouldn't return `{}` on cache miss.  If there is nothing in the cache, and no labels on the pod, we still need to write the endpoint.